src/sentry/rules/processing/delayed_processing.py

```diff
@@ -200,6 +202,31 @@ def parse_rulegroup_to_event_data(
     return parsed_rulegroup_to_event_data
 
 
+def bulk_fetch_events(event_ids: list[str], project_id: int) -> set[Event]:
+    node_id_to_event_id = {
+        Event.generate_node_id(project_id, event_id=event_id): event_id for event_id in event_ids
+    }
+
+    try:
+        bulk_data = nodestore.get_multi(list(node_id_to_event_id.keys()))
+    except (ServiceUnavailable, DeadlineExceeded):
+        try:
+            bulk_data = nodestore.get_multi(list(node_id_to_event_id.keys()))
+        except (ServiceUnavailable, DeadlineExceeded):
+            bulk_data = nodestore.get_multi(list(node_id_to_event_id.keys()))
+
+    bulk_events = set()
+    for node_id, data in bulk_data.items():
+        if node_id in node_id_to_event_id:
+            event_id = node_id_to_event_id[node_id]
+            if data is not None:
+                event = Event(event_id=event_id, project_id=project_id)
+                event.data = data
+            bulk_events.add(event)
```

Comment:
Rather than nesting the if statements we can pull them to the top level.
```suggestion
        event_id = node_id_event_id.get(node_id, None)
        if data is not None and event_id is not None:
            event = Event(event_id=event_id, project_id=project_id)
            event.data = data
            bulk_events.add(event)
```

This refactor shows an interesting side effect of your current code, that you could have an event_id but not data - at least in the logic. It's okay because the set will ensure uniqueness and not add the event multiple times, but it could be confusing in the future. It'd be better to only add the event to bulk events if the event is created though.